src/main/groovy/com/blackducksoftware/integration/hub/detect/bomtool/npm/NpmNodeModulesDependencyFinder.groovy

```diff
@@ -0,0 +1,154 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.blackducksoftware.integration.hub.detect.bomtool.npm
+
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Component
+
+import com.blackducksoftware.integration.hub.bdio.simple.model.DependencyNode
+import com.blackducksoftware.integration.hub.bdio.simple.model.Forge
+import com.blackducksoftware.integration.hub.bdio.simple.model.externalid.NameVersionExternalId
+import com.blackducksoftware.integration.hub.detect.util.FileFinder
+import com.google.gson.Gson
+import com.google.gson.stream.JsonReader
+
+@Component
+class NpmNodeModulesDependencyFinder {
+
+	private final Logger logger = LoggerFactory.getLogger(NpmNodeModulesDependencyFinder.class)
+
+	@Autowired
+	Gson gson
+
+	@Autowired
+	FileFinder fileFinder
+
+	/*
+	 * arraylist used to check for circular dependencies
+	 */
+	def modules = new HashSet<String>()
+
+	public DependencyNode generateDependencyNode(String nodeProjectDirectory) {
+		def packageJsonFile = fileFinder.findFile(nodeProjectDirectory, NpmConstants.PACKAGE_JSON)
+		def nodeModulesFile = fileFinder.findFile(nodeProjectDirectory, NpmConstants.NODE_MODULES)
+
+		PackageNode pNode
+
+		if(packageJsonFile?.exists()) {
+			pNode = deserializeJsonToPackageNode(packageJsonFile)
+		}
+
+		if(nodeModulesFile.exists() && nodeModulesFile.isDirectory()) {
+			if(pNode) {
+				logger.info("Traversing project directory starting from package.json dependencies")
+				return createNodeFromPackageNode(pNode, nodeModulesFile)
+			} else {
+				logger.info("No package.json found, parsing node_modules folder")
+				return createNodeFromNodeModules(nodeModulesFile)
+			}
+		} else if(pNode) {
+			/*
+			 * We have only the package.json and no node_modules folder which would require REST calls to complete.
+			 * This would be a good place to check for shrinkwrap/package-lock files
+			 */
+			logger.error("No node_modules folder located in target path. Can't run program")
+		}
+	}
+
+	/*
+	 * The best case scenario for parsing npm without npm installed (uses package.json and node_modules).
+	 * Note: Looks to be faster than the Cli finder
+	 */
+	public DependencyNode createNodeFromPackageNode(PackageNode pNode, File nodeModulesFile) {
+		DependencyNode node = createDependencyNode(pNode)
+
+		//Had to create a HashSet object before adding content because of issues with the keySet method for maps
+		def allDependencies = new HashSet<String>()
+		allDependencies.addAll(pNode?.dependencies?.keySet())
+		Set<String> optionalDependencies = pNode?.optionalDependencies?.keySet()
+
+		if(optionalDependencies) {
+			allDependencies.addAll(optionalDependencies)
+		}
+
+		for(String filename : allDependencies) {
+			if(modules.contains(filename)) {
```

Comment:
modules is a HashSet<String>, it seems this is only being used as a way to ensure we only add the node once. So modules should be changes to a HashMap<String, String> it will make the lookup time faster